File: DiscordConnector/PublicAPI/api/v0/category/create.py
from fastapi import APIRouter, Depends
from DiscordConnector.PublicAPI.dependencies import get_category_service
from ..schemas import CategoryCreate, CategoryResponse

router = APIRouter()


@router.post("/create", response_model=CategoryResponse)
async def create_category(
    data: CategoryCreate,
    service=Depends(get_category_service),
):
    category = await service.create_category(data.name, data.position)
    return CategoryResponse(
        id=str(category.id),
        name=category.name,
        position=category.position,
    )

File: DiscordConnector/src/PublicAPI/api/v0/category/create.ts
import { createRoute, type OpenAPIHono } from "@hono/zod-openapi";
import { categoryResponse } from "../../../responses";
import {
  bearerSecurity,
  CategorySchema,
  commonErrorResponses,
  createCategoryBodySchema,
  getServices,
  jsonBody,
  jsonContent,
  requireRoleMiddleware,
  type AppEnv,
} from "../../../shared";

export function registerCreateCategoryRoute(app: OpenAPIHono<AppEnv>): void {
  app.openapi(
    createRoute({
      method: "post",
      path: "/api/v0/category/create",
      security: bearerSecurity,
      middleware: requireRoleMiddleware("admin"),
      request: {
        body: jsonBody(createCategoryBodySchema, "Category creation payload"),
      },
      responses: {
        200: jsonContent(CategorySchema, "Created category"),
        ...commonErrorResponses(),
      },
    }),
    async (c) => {
      const body = c.req.valid("json");
      const category = await getServices(c.env).categoryService.createCategory(
        body.name,
        body.position ?? null,
      );
      return c.json(categoryResponse(category), 200);
    },
  );
}
